libs/header/p2p/session.go

```diff
@@ -139,30 +145,47 @@ func (s *session[H]) doRequest(
 
 	r, size, duration, err := sendMessage(ctx, s.host, stat.peerID, s.protocolID, req)
 	if err != nil {
-		log.Errorw("requesting headers from peer failed", "failed peer", stat.peerID, "err", err)
-		select {
-		case <-s.ctx.Done():
-		case s.reqCh <- req:
-			stat.decreaseScore()
-			log.Debug("retrying the request from different peer")
-		}
-		return
+		// we should not punish peer at this point and should try to parse responses, despite that error was received.
+		log.Debugw("requesting headers from peer failed", "failed peer", stat.peerID, "err", err)
 	}
 
 	h, err := s.processResponse(r)
 	if err != nil {
 		switch err {
 		case header.ErrNotFound:
+			// errNoResponses means that we did not fetch any headers from the peer, so we can decrease its score.
+		case errEmptyResponse:
+			stat.decreaseScore()
 		default:
 			s.peerTracker.blockPeer(stat.peerID, err)
+			log.Errorw("parsing headers failed", "failed peer", stat.peerID, "err", err)
 		}
 		select {
 		case <-s.ctx.Done():
 		case s.reqCh <- req:
 		}
 		return
 	}
-	log.Debugw("request headers from peer succeeded", "peer", stat.peerID, "amount", req.Amount)
+
+	log.Debugw("request headers from peer succeeded",
+		"peer", stat.peerID,
+		"receivedAmount", len(h),
+		"requestedAmount", req.Amount,
+	)
+
+	// ensure that we received the correct amount of headers.
+	if uint64(len(h)) != req.Amount {
+		from := uint64(h[len(h)-1].Height())
+		amount := req.Amount - from
+		select {
+		case <-s.ctx.Done():
```

Comment:
return here, no?

Reply:
it does. Next line is another case

Reply:
`updateStats` + return bc we are receiving ctx.Done after external calls happened.

Reply:
Okay i could swear there was no `return` statement here when i reviewed.